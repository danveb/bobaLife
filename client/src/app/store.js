import { combineReducers, configureStore } from "@reduxjs/toolkit"; 
import authReducer from "../redux/auth/authSlice"; 
import productsReducer from "../redux/products/productsSlice"; 
import cartReducer from "../redux/cart/cartSlice"; 
import {
    persistStore,
    persistReducer,
    FLUSH,
    REHYDRATE,
    PAUSE,
    PERSIST,
    PURGE,
    REGISTER,
} from 'redux-persist';
import storage from "redux-persist/lib/storage"; 

// Perhaps combine all reducers and persist? 
// even persist user so we don't save user data in localStorage?
// combineReducers from rootReducer
const rootReducer = combineReducers({
    auth: authReducer, 
    products: productsReducer, 
    cart: cartReducer, 
}); 

const persistConfig = {
    key: 'root',
    storage,
}
  
// const persistedReducer = persistReducer(persistConfig, cartReducer)
const persistedReducer = persistReducer(persistConfig, rootReducer)

export const store = configureStore({
    // reducer: {
    //     auth: authReducer, 
    //     products: productsReducer, 
    //     cart: persistedReducer, 
    // },
    reducer: persistedReducer, 
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
            },
        }),
}); 

export const persistor = persistStore(store) 