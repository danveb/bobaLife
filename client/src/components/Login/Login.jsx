import { useEffect, useState } from "react"; 
import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { toast } from "react-toastify"; 
import { login, reset } from "../../redux/auth/authSlice"; 
import { Spinner } from "../index";
import { LoginContainer, Wrapper, Main, H1, P, Form, Label, Input, Btn } from "../../styles/Login.style";

const Login = () => {
    // useState 
    const [formData, setFormData] = useState({
        email: "", 
        password: "",
    }); 

    // destructured form data
    const { email, password } = formData; 

    // useNavigate 
    const navigate = useNavigate(); 

    // useDispatch 
    const dispatch = useDispatch(); 

    // select from state.auth
    const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth); 

    // useEffect
    useEffect(() => {
        if(isError) {
            toast.error(message, {
                position: "top-center", 
                autoClose: 2500, 
                pauseOnHover: true, 
            }); 
        }; 
        if(isSuccess) {
            toast.success(`Welcome back ${user.username}`, {
                position: "top-center", 
                autoClose: 1500, 
                pauseOnHover: true, 
            }); 
            navigate("/"); 
        }; 
        // redirect existing user back to home
        if(user) {
            navigate("/"); 
        }
        dispatch(reset()); 
    }, [user, isError, isSuccess, message, navigate, dispatch]);


    // handleChange
    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState, 
            [e.target.name]: e.target.value
        })); 
    }; 

    // handleSubmit
    const handleSubmit = (e) => {
        e.preventDefault(); 
        const userData = {
            email, 
            password
        };
        // error check with toast 
        if(email.length === 0) {
            toast.error("Email field is empty. Please enter it again.", {
                position: "top-center", 
                autoClose: 1500, 
                pauseOnHover: true, 
            }); 
        } else if(password.length === 0) {
            toast.error("Password field is empty. Please enter it again.", {
                position: "top-center", 
                autoClose: 1500, 
                pauseOnHover: true, 
            }); 
        } else {
            dispatch(login(userData)); 
        }
    }; 

    if(isLoading) {
        return <Spinner />
    };

    return (
        <LoginContainer>
            <Wrapper>
                <Main>
                    <H1>Sign In</H1>
                    <P>Enter your account <span>credentials</span></P>
                </Main>
                <Form onSubmit={handleSubmit}>
                    <Label
                        htmlFor="email"
                    >
                        Email</Label>
                    <Input 
                        id="email"
                        name="email"
                        value={email}
                        onChange={handleChange}
                        type="email" 
                        placeholder="Your email" 
                    />
                    <Label
                        htmlFor="password"
                    >
                        Password</Label>
                    <Input 
                        id="password"
                        name="password"
                        value={password}
                        onChange={handleChange}
                        type="password" 
                        placeholder="Your password" 
                    />
                    <Btn>Sign In</Btn>
                    <Link to="/register">Need a new account?</Link>
                </Form>
            </Wrapper>
        </LoginContainer>
    )
    
}

export default Login