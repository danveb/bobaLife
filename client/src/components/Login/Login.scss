.login {
    display: flex; 
    justify-content: center; 
    align-items: center; 
    height: calc(100vh - 79px); 

    @media screen and (max-width: 768px) {
        height: 70vh; 
    }

    .login-wrapper {
        width: 40%; 
        padding: 15px; 

        @media screen and(max-width: 1024px) {
            width: 50%; 
        }

        @media screen and(max-width: 768px) {
            width: 80%; 
        }

        @media screen and(max-width: 480px) {
            width: 90%; 
        }

        h1 {
            font-size: 24px; 
            font-weight: 300; 
        }

        p {
            font-size: 18px; 

            @media screen and (max-width: 768px) {
                display: none; 
            }
        }

        form {
            display: flex; 
            flex-direction: column; 
            font-family: "Quicksand", sans-serif; 

            input {
                flex: 1; 
                min-width: 40%; 
                margin: 10px 0px; 
                padding: 10px; 
                font-family: inherit; 
            }

            button {
                width: 100%; 
                border: none; 
                padding: 15px 20px; 
                background-color: #000; 
                color: #fff; 
                font-weight: 300; 
                cursor: pointer; 
                margin: 10px 0px; 
                font-family: inherit; 
            }

            a {
                color: #000; 
                text-decoration: underline; 
                text-align: right; 
                margin: 10px 0px; 
                cursor: pointer; 
                font-family: inherit; 
            }
        }
    }
}