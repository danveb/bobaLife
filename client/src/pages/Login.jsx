import { useEffect, useState } from "react"; 
import { Link, useNavigate } from "react-router-dom"; 
import { useSelector, useDispatch } from "react-redux"; 
import { toast } from "react-toastify"; 
import { login, reset } from "../redux/auth/authSlice"; 
import { Navbar, Menu, Spinner } from "../components";
import "./Login.scss"; 

const Login = ({ menuOpen, setMenuOpen }) => {
    // useState 
    const [formData, setFormData] = useState({
        email: "", 
        password: "",
    }); 

    // destructured form data
    const { email, password } = formData; 

    // useNavigate 
    const navigate = useNavigate(); 

    // useDispatch 
    const dispatch = useDispatch(); 

    // select from state.auth
    const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth); 

    // useEffect
    useEffect(() => {
        if(isError) {
            toast.error(message); 
        }; 

        if(isSuccess || user) {
            navigate("/"); 
        }; 
        dispatch(reset()); 
    }, [user, isError, isSuccess, message, navigate, dispatch]);


    // handleChange
    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState, 
            [e.target.name]: e.target.value
        })); 
    }; 

    // handleSubmit
    const handleSubmit = (e) => {
        e.preventDefault(); 
        const userData = {
            email, 
            password
        };
        dispatch(login(userData)); 
    }; 

    if(isLoading) {
        return <Spinner />
    }

    return (
        <div>
            <Navbar menuOpen={menuOpen} setMenuOpen={setMenuOpen} />
            <Menu menuOpen={menuOpen} setMenuOpen={setMenuOpen} />
            <div className="login">
                <div className="login-wrapper">
                    <h1>Sign In</h1>
                    <p>Please enter your credentials</p>
                    <form onSubmit={handleSubmit} >
                        <label htmlFor="email" />
                        <input
                            id="email" 
                            type="email"
                            name="email"
                            placeholder="Enter your email"
                            value={email} 
                            onChange={handleChange}
                        />

                        <label htmlFor="password" />
                        <input 
                            id="password"
                            type="password"
                            name="password"
                            placeholder="Enter your password"
                            value={password} 
                            onChange={handleChange}
                        />
                        <button type="submit">Sign In</button>
                        <Link to="/register">Register New Account</Link>
                    </form>
                </div>
            </div>
        </div>
    )
    
}

export default Login