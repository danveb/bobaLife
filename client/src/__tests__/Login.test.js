import { fireEvent, render, screen } from "@testing-library/react"; 
import { store } from "../app/store"; 
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import Login from "../components/Login/Login"; 

describe("Login component", () => {
    // smoke test
    test("renders Login component without crashing", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
    }); 

    // snapshot test
    test("matches snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        expect(asFragment()).toMatchSnapshot(); 
    });

    // getByText
    test("should display header display", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        const paragraphTag = screen.getByText("Enter your account", { exact: true }); 
        expect(paragraphTag).toBeInTheDocument(); 
    });

    // getByLabelText
    test("should find email label", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        );
        const label = screen.getByLabelText("Email"); 
        expect(label).toBeInTheDocument(); 
    }); 

    // getByLabelText
    test("email input field accepts correct value", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        const emailInput = screen.getByLabelText("Email"); 
        expect(emailInput.value).toMatch(""); 
        fireEvent.change(emailInput, { target: { value: "hello@gmail.com" }}); 
        expect(emailInput.value).toMatch("hello@gmail.com"); 
    }); 

    // getByLabelText
    test("input email triggers error for failing required data", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        const emailInputNode = screen.getByLabelText("Email"); 
        expect(emailInputNode.value).toMatch(""); 
        fireEvent.change(emailInputNode, { target: { value: "hello" }}); 
        expect(emailInputNode.value).not.toMatch("hello@gmail.com"); 
    }); 

    // getByLabelText
    test("password input field accepts correct value", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        const passwordInputNode = screen.getByLabelText("Password"); 
        expect(passwordInputNode.value).toBe(""); 
        fireEvent.change(passwordInputNode, { target: { value: "helloworld" }}); 
        expect(passwordInputNode.value).toBe("helloworld"); 
    });

    // getByText
    test("displays link to navigate to register page", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        ); 
        const newAccountText = screen.getByText("Need a new account?"); 
        expect(newAccountText).toBeInTheDocument(); 
    }); 

    // getByLabelText
    test("displays initial state for username/password", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        );
        const emailInput = screen.getByLabelText("Email");
        const passwordInput = screen.getByLabelText("Password");
        expect(emailInput.value).toBe(""); 
        expect(passwordInput.value).toBe(""); 
    });

    // mock login functionality
    test("mock login functionality", () => {
        const email = "jojo@gmail.com"; 
        const password = "jojo"; 
        const mockSubmit = jest.fn();

        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login onSubmit={mockSubmit(email, password)}/>
                </BrowserRouter>
            </Provider>
        );
        const emailInput = screen.getByLabelText("Email"); 
        const passwordInput = screen.getByLabelText("Password"); 
        const signInBtn = screen.getByRole("button");

        fireEvent.change(emailInput, { target: { value: email }}); 
        fireEvent.change(passwordInput, { target: { value: password }}); 
        fireEvent.click(signInBtn); 
        expect(mockSubmit).toHaveBeenCalled(); 
    }); 
});