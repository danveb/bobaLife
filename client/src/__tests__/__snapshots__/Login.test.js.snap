// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-caXVBt gJpVKA"
  >
    <div
      class="sc-hjriPb dWFijc"
    >
      <div
        class="sc-gUAEMC eoOTrv"
      >
        <h1
          class="sc-fWjsSh kVywhz"
        >
          Sign In
        </h1>
        <p
          class="sc-fHsOPI kieKWP"
        >
          Enter your account 
          <span>
            credentials
          </span>
        </p>
      </div>
      <form
        class="sc-gzzPqb gPOFHw"
      >
        <label
          class="sc-gkJlnC juEHOj"
          for="email"
        >
          Email
        </label>
        <input
          class="sc-fmrZth gfnrlj"
          id="email"
          name="email"
          placeholder="Your email"
          type="email"
          value=""
        />
        <label
          class="sc-gkJlnC juEHOj"
          for="password"
        >
          Password
        </label>
        <input
          class="sc-fmrZth gfnrlj"
          id="password"
          name="password"
          placeholder="Your password"
          type="password"
          value=""
        />
        <button
          class="sc-eXBvqI bsbuvm"
        >
          Sign In
        </button>
        <a
          href="/register"
        >
          Need a new account?
        </a>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
